File: corpus/generate_pdf.js
/**
 * generate_pdf.js — Phase 1 helper
 * Converts research_degrees_handbook_source.md → research_degrees_handbook.pdf
 * Uses: pdf-lib  (npm install pdf-lib)
 * Run from: corpus/
 *   node generate_pdf.js
 */

const { execSync } = require('child_process');
const fs = require('fs');
const path = require('path');

// Install pdf-lib if not present
function ensureLib() {
  try { require.resolve('pdf-lib'); }
  catch {
    console.log('Installing pdf-lib...');
    execSync('npm install pdf-lib --no-save', { stdio: 'inherit', cwd: __dirname });
  }
}

ensureLib();

const { PDFDocument, rgb, StandardFonts } = require('pdf-lib');

async function main() {
  const srcPath = path.join(__dirname, 'research_degrees_handbook_source.md');
  const outPath = path.join(__dirname, 'research_degrees_handbook.pdf');

  const mdText = fs.readFileSync(srcPath, 'utf-8');
  const lines = mdText.split(/\r?\n/);

  const doc = await PDFDocument.create();
  const fontBold = await doc.embedFont(StandardFonts.HelveticaBold);
  const fontReg  = await doc.embedFont(StandardFonts.Helvetica);

  const PAGE_W = 595;
  const PAGE_H = 842;
  const ML = 50, MR = 50, MT = 50, MB = 50;
  const contentW = PAGE_W - ML - MR;

  let page = doc.addPage([PAGE_W, PAGE_H]);
  let y = PAGE_H - MT;
  let pageNum = 1;

  function newPage() {
    // add page number to current page
    page.drawText(`${pageNum}`, {
      x: PAGE_W / 2 - 5, y: 20,
      size: 8, font: fontReg, color: rgb(0.5, 0.5, 0.5)
    });
    page = doc.addPage([PAGE_W, PAGE_H]);
    y = PAGE_H - MT;
    pageNum++;
  }

  function ensureSpace(needed) {
    if (y - needed < MB) newPage();
  }

  function drawWrapped(text, font, size, color, indent = 0, lineGap = 2) {
    const maxW = contentW - indent;
    const words = text.split(' ');
    let currentLine = '';

    for (const word of words) {
      const test = currentLine ? `${currentLine} ${word}` : word;
      const w = font.widthOfTextAtSize(test, size);
      if (w > maxW && currentLine) {
        ensureSpace(size + lineGap);
        page.drawText(currentLine, {
          x: ML + indent, y,
          size, font, color
        });
        y -= (size + lineGap);
        currentLine = word;
      } else {
        currentLine = test;
      }
    }
    if (currentLine) {
      ensureSpace(size + lineGap);
      page.drawText(currentLine, {
        x: ML + indent, y,
        size, font, color
      });
      y -= (size + lineGap);
    }
  }

  const BLACK   = rgb(0.1, 0.1, 0.1);
  const DARKBLUE = rgb(0.08, 0.16, 0.35);
  const MEDBLUE  = rgb(0.12, 0.24, 0.47);
  const LITBLUE  = rgb(0.20, 0.31, 0.55);

  for (const rawLine of lines) {
    const line = rawLine.trim();

    if (line === '') {
      y -= 5;
      continue;
    }

    if (line === '---') {
      ensureSpace(10);
      page.drawLine({
        start: { x: ML, y },
        end:   { x: PAGE_W - MR, y },
        thickness: 0.5,
        color: rgb(0.7, 0.7, 0.7)
      });
      y -= 8;
      continue;
    }

    // H1
    if (/^# (?!#)/.test(rawLine)) {
      const text = line.replace(/^# /, '');
      ensureSpace(22);
      y -= 4;
      drawWrapped(text, fontBold, 14, DARKBLUE);
      y -= 4;
      continue;
    }

    // H2
    if (/^## (?!#)/.test(rawLine)) {
      const text = line.replace(/^## /, '');
      ensureSpace(18);
      y -= 3;
      drawWrapped(text, fontBold, 12, MEDBLUE);
      y -= 3;
      continue;
    }

    // H3
    if (/^### /.test(rawLine)) {
      const text = line.replace(/^### /, '');
      ensureSpace(15);
      y -= 2;
      drawWrapped(text, fontBold, 10.5, LITBLUE);
      y -= 2;
      continue;
    }

    // Bullet
    if (/^- /.test(rawLine)) {
      const text = line.replace(/^- /, '');
      const clean = text.replace(/\*\*(.+?)\*\*/g, '$1').replace(/\*(.+?)\*/g, '$1').replace(/`(.+?)`/g, '$1');
      drawWrapped('\u2022  ' + clean, fontReg, 9.5, BLACK, 10);
      y -= 2;
      continue;
    }

    // Table rows
    if (/^\|/.test(line)) {
      if (/^\|[-|]+\|$/.test(line)) { y -= 2; continue; }
      const cols = line.split('|').slice(1, -1).map(c => c.trim());
      const colText = cols.join('  |  ');
      const clean = colText.replace(/\*\*(.+?)\*\*/g, '$1');
      drawWrapped(clean, fontReg, 8.5, BLACK);
      y -= 1;
      continue;
    }

    // Bold-only line
    if (/^\*\*(.+)\*\*$/.test(line)) {
      const text = line.replace(/^\*\*/, '').replace(/\*\*$/, '');
      drawWrapped(text, fontBold, 9.5, BLACK);
      y -= 2;
      continue;
    }

    // body
    const clean = line.replace(/\*\*(.+?)\*\*/g, '$1').replace(/\*(.+?)\*/g, '$1').replace(/`(.+?)`/g, '$1');
    drawWrapped(clean, fontReg, 9.5, BLACK);
    y -= 1;
  }

  // Final page number
  page.drawText(`${pageNum}`, {
    x: PAGE_W / 2 - 5, y: 20,
    size: 8, font: fontReg, color: rgb(0.5, 0.5, 0.5)
  });

  const bytes = await doc.save();
  fs.writeFileSync(outPath, bytes);
  const stats = fs.statSync(outPath);
  console.log(`PDF generated: ${outPath} (${(stats.size / 1024).toFixed(1)} KB, ${doc.getPageCount()} pages)`);
}

main().catch(err => { console.error(err); process.exit(1); });

File: corpus/generate_pdf.py
"""
generate_pdf.py — Phase 1 helper script.
Converts research_degrees_handbook_source.md → research_degrees_handbook.pdf
Uses fpdf2 (pip install fpdf2).
Run from: corpus/
"""

import sys
import re
from pathlib import Path


def install_and_import():
    import subprocess
    subprocess.check_call([sys.executable, "-m", "pip", "install", "fpdf2", "--quiet"])
    from fpdf import FPDF
    return FPDF


def parse_markdown_lines(md_text: str):
    """
    Yields (style, text) tuples where style is one of:
      'h1', 'h2', 'h3', 'bold', 'body', 'blank', 'hr'
    """
    for line in md_text.splitlines():
        stripped = line.rstrip()
        if stripped.startswith("# ") and not stripped.startswith("## "):
            yield ("h1", stripped[2:])
        elif stripped.startswith("## ") and not stripped.startswith("### "):
            yield ("h2", stripped[3:])
        elif stripped.startswith("### "):
            yield ("h3", stripped[4:])
        elif stripped.startswith("---"):
            yield ("hr", "")
        elif stripped == "":
            yield ("blank", "")
        elif re.match(r"^\*\*(.+)\*\*$", stripped):
            yield ("bold", re.match(r"^\*\*(.+)\*\*$", stripped).group(1))
        else:
            # strip inline markdown bold/italic for plain PDF
            clean = re.sub(r"\*\*(.+?)\*\*", r"\1", stripped)
            clean = re.sub(r"\*(.+?)\*", r"\1", clean)
            clean = re.sub(r"`(.+?)`", r"\1", clean)
            yield ("body", clean)


def build_pdf(source_path: Path, output_path: Path):
    try:
        from fpdf import FPDF
    except ImportError:
        FPDF = install_and_import()

    pdf = FPDF()
    pdf.set_auto_page_break(auto=True, margin=20)
    pdf.add_page()
    pdf.set_margins(20, 20, 20)

    md_text = source_path.read_text(encoding="utf-8")

    for style, text in parse_markdown_lines(md_text):
        if style == "h1":
            pdf.set_font("Helvetica", "B", 16)
            pdf.set_text_color(20, 40, 80)
            pdf.ln(4)
            pdf.multi_cell(0, 8, text)
            pdf.ln(3)
        elif style == "h2":
            pdf.set_font("Helvetica", "B", 13)
            pdf.set_text_color(30, 60, 120)
            pdf.ln(3)
            pdf.multi_cell(0, 7, text)
            pdf.ln(2)
        elif style == "h3":
            pdf.set_font("Helvetica", "B", 11)
            pdf.set_text_color(50, 80, 140)
            pdf.ln(2)
            pdf.multi_cell(0, 6, text)
            pdf.ln(1)
        elif style == "hr":
            pdf.set_draw_color(180, 180, 180)
            pdf.ln(2)
            pdf.line(pdf.get_x(), pdf.get_y(), pdf.get_x() + 170, pdf.get_y())
            pdf.ln(3)
        elif style == "blank":
            pdf.ln(3)
        elif style == "bold":
            pdf.set_font("Helvetica", "B", 10)
            pdf.set_text_color(0, 0, 0)
            pdf.multi_cell(0, 5, text)
            pdf.ln(1)
        else:  # body
            pdf.set_font("Helvetica", "", 10)
            pdf.set_text_color(30, 30, 30)
            # handle bullet points
            if text.startswith("- "):
                pdf.set_x(pdf.get_x() + 5)
                pdf.multi_cell(0, 5, "\u2022 " + text[2:])
            else:
                pdf.multi_cell(0, 5, text)

    pdf.output(str(output_path))
    print(f"PDF generated: {output_path} ({output_path.stat().st_size:,} bytes)")


if __name__ == "__main__":
    corpus_dir = Path(__file__).parent
    source = corpus_dir / "research_degrees_handbook_source.md"
    output = corpus_dir / "research_degrees_handbook.pdf"

    if not source.exists():
        print(f"ERROR: Source not found: {source}")
        sys.exit(1)

    build_pdf(source, output)
